File: Leetcode/0001-two-sum/0001-two-sum.java
//시간복잡도: O(n)
//공간복잡도: O(1)
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> positions = new HashMap<>();

        for(int i = 0; i < nums.length; i++) {
            positions.put(nums[i], i);
        }

        for(int i = 0; i < nums.length; i++) {
            if(!positions.containsKey(target - nums[i])) {
                continue;
            }

            int position = positions.get(target - nums[i]);
            
            if(position != i) {
                return new int[]{i, position};
            }
        }

        return new int[]{};
    }
}

File: Leetcode/0001-two-sum/0001-two-sum.kt
class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val numsMap = mutableMapOf<Int, Int>()

        for ((i, num) in nums.withIndex()) {
            if(numsMap.containsKey(target - num)) {
                return intArrayOf(i, numsMap[target - num] ?: 0)
            }

            numsMap[num] = i
        }

        return intArrayOf(0, 0)
    }
}